Lute/reification step2

lute47lillo opened the pull request:
Draft PR. No changes have yet been made. Addressing issue #32.

lute47lillo commented on src/eclingo/parsing/transformers/ast_reify.py:
```diff
@@ -0,0 +1,11 @@
+
+def symbolic_literal_to_term(x: AST) -> AST:
+    assert x.ast_type == ASTType.Literal and x.atom.ast_type == ASTType.SymbolicAtom
+    arg = ast.SymbolicTerm(x.location, clingo.Function('b', [], True))
```
Is there a reason for using 'b' and "a" when creating the SymbolicTerm and Function as placeholders?

lute47lillo replied:
So, before I start looking into how to reify a Literal into a Function, I just wanted to know that this code above is arbitry right?

jorgefandinno replied:
'a' and 'b' are just the names.

This is specific code to pass exactly one test and check that the test make sense.

It does not transform ```x``` into anything. It just creates a new term that on purpose coincide with the first test. Obviously, any other test will fail.

lute47lillo commented:
@jorgefandinno Have you deleted jorge/filter_body?
https://github.com/jorgefandinno/eclingo/blob/3a5446b44024309ca41c9efae39d3fd1b722ccf3/requirements.txt#L5
Because I'm getting errors now since nox it is now not matching this branch when running requirements.

lute47lillo commented:
_not1(a) != not1(a)
{'arguments': [], 'ast_type': 'Function', 'external': 0, 'name': 'not1(a)'}
{'arguments': [{'ast_type': 'SymbolicTerm', 'symbol': 'a'}],
 'ast_type': 'Function',
 'external': 0,
 'name': 'not1'}_

I have just done a first iteration on what it would be, I was just moving things around to see how I could access all the elements and reify them. However, I have one question:
The code above is the error that I get now, for clarity to continue on it tomorrow, when I convert a Literal into a Function, do I want the arguments to be like the second declaration in the error, right? Therefore, reifying only the name to be not1, and leaving the symbol: "a" as part of the argument?

I'll continue on it tomorrow

lute47lillo commented:
> _not1(a) != not1(a) {'arguments': [], 'ast_type': 'Function', 'external': 0, 'name': 'not1(a)'} {'arguments': [{'ast_type': 'SymbolicTerm', 'symbol': 'a'}], 'ast_type': 'Function', 'external': 0, 'name': 'not1'}_
> 
> I have just done a first iteration on what it would be, I was just moving things around to see how I could access all the elements and reify them. However, I have one question: The code above is the error that I get now, for clarity to continue on it tomorrow, when I convert a Literal into a Function, do I want the arguments to be like the second declaration in the error, right? Therefore, reifying only the name to be not1, and leaving the symbol: "a" as part of the argument?
> 
> I'll continue on it tomorrow

Nvm about this. I got it almost working for all positive terms. I'll deal now with negations of type (-a) -> (-a). 
not and not not negations are working.

lute47lillo commented:
@jorgefandinno Do we want UnaryOperators to be also transformed into ASTType.Functions?
Because what the assert is expecting is a term of ASTType.UnaryOperator once is reified, those are the ones I have commented out, since I don't know if I should construct a Function from a Literal, like you said, making the test to fail. Or if I should construct an UnaryOperator

<img width="438" alt="Screen Shot 2022-07-08 at 12 26 50 PM" src="https://user-images.githubusercontent.com/65786380/178041132-0d6d423c-5b9c-4f59-bbe1-6ac05589b0b0.png">

jorgefandinno commented:
I fixed the problem with locations and reformatted the code using black.
For instructions on how to use it, see https://github.com/jorgefandinno/eclingo/blob/develop/README.md.
Let me know if you have any doubt.

I added further test for reify with variables and these fail.

@lute47lillo  Can you check?

lute47lillo commented:
> I fixed the problem with locations and reformatted the code using black. For instructions on how to use it, see https://github.com/jorgefandinno/eclingo/blob/develop/README.md. Let me know if you have any doubt.
> 
> I added further test for reify with variables and these fail.
> 
> @lute47lillo Can you check?

@jorgefandinno Sounds good. I'll get into it later today. Thank you!

lute47lillo commented:
I have done some updates but I still have to go over some cases, however I don't know if such case could exist. I was thinking on:
a(b(t(X,Z)),c,Y) -> There is a Function within a Function with 2 Variables. I'll have to iterate through them. Does that possibility of having a literal b with an argument that is another literal with 2 variables? Or that is not going to happen?

And then the possibility of having a(b(X,Z),c,Y) is also possible? where there is a function with 2 variables right?

In that case I have to handle recursive calls. Let me know if those are possible scenarios and I'll refactor the code

@jorgefandinno

jorgefandinno commented:
> I have done some updates but I still have to go over some cases, however I don't know if such case could exist. I was thinking on: a(b(t(X,Z)),c,Y) -> There is a Function within a Function with 2 Variables. I'll have to iterate through them. Does that possibility of having a literal b with an argument that is another literal with 2 variables? Or that is not going to happen?
> 
> And then the possibility of having a(b(X,Z),c,Y) is also possible? where there is a function with 2 variables right?
> 
> In that case I have to handle recursive calls. Let me know if those are possible scenarios and I'll refactor the code
> 
> @jorgefandinno

It is possible to have functions as you describe. In fact as nested as you want.
But you should not need to do anything beyond the first level because those are already terms. Only the ```a(...)``` is a literal and has to become a term. ```b(t(X,Z))``` should be already a term.

lute47lillo commented:
> > I have done some updates but I still have to go over some cases, however I don't know if such case could exist. I was thinking on: a(b(t(X,Z)),c,Y) -> There is a Function within a Function with 2 Variables. I'll have to iterate through them. Does that possibility of having a literal b with an argument that is another literal with 2 variables? Or that is not going to happen?
> > And then the possibility of having a(b(X,Z),c,Y) is also possible? where there is a function with 2 variables right?
> > In that case I have to handle recursive calls. Let me know if those are possible scenarios and I'll refactor the code
> > @jorgefandinno
> 
> It is possible to have functions as you describe. In fact as nested as you want. But you should not need to do anything beyond the first level because those are already terms. Only the `a(...)` is a literal and has to become a term. `b(t(X,Z))` should be already a term.

Okay, I have fixed that for now. I have added a new test for coverage for the case of multiple variables. I made the pull request also available for code review now.

jorgefandinno commented:
I refactor the code to simplify the implementation.

Please check.

lute47lillo commented:
> I refactor the code to simplify the implementation.
> 
> Please check.

Yes, I've seen them. they work on my end. 
Way more simple code. I'll try to make my code simpler on next issues.

I think this pull request can be merged into the branch reification if that's all the work needed 👍🏻
